mscore/timeline.cpp

```diff
@@ -0,0 +1,1514 @@
+            non_visible->setZValue(-3);
+
+            QGraphicsPathItem* visible = new QGraphicsPathItem(pp.simplified());
+            visible->setPen(non_visible_p);
+            visible->setBrush(Qt::NoBrush);
+            visible->setZValue(-2);
+
+            //Find old path, remove it
+            for (QGraphicsItem* gi : scene()->items()) {
+                  if (gi->type() == QGraphicsPathItem().type()) {
+                        QGraphicsPathItem* p = static_cast<QGraphicsPathItem*>(gi);
+                        QBrush b1 = p->brush();
+                        QPen p1 = p->pen();
+                        if (b1 == non_visible_b || p1 == non_visible_p)
+                              scene()->removeItem(gi);
+                        }
+
+                  }
+
+            scene()->addItem(non_visible);
+            scene()->addItem(visible);
+            }
+      else {
+            qDebug() << "No scoreview found.";
+            }
+
+      }
+
+
+
+//---------------------------------------------------------
+//   nstaves
+//---------------------------------------------------------
+
+int Timeline::nstaves()
+      {
+      QList<Staff*> list = _score->staves();
+      int total = 0;
+      for (Staff* s : list) {
+            if (s->show())
+                  total++;
+            }
+      return total;
+      }
+
+//---------------------------------------------------------
+//   colorBox
+//---------------------------------------------------------
+
+QColor Timeline::colorBox(QGraphicsRectItem* item)
+      {
+      Measure* meas = static_cast<Measure*>(item->data(2).value<void*>());
+      int stave = item->data(0).value<int>();
+      for (Segment* s = meas->first(); s; s = s->next()) {
+            if (s->segmentType() != SegmentType::ChordRest) {
+                  continue;
+                  }
+            for (int i = stave * VOICES; i < stave * VOICES + VOICES; i++) {
+                  ChordRest* cr = s->cr(i);
+                  if (cr) {
+                        if (cr->type() == ElementType::CHORD) {
+                              return QColor(Qt::gray);
+                              }
+                        }
+                  }
+            }
+      return QColor(211,211,211);
+      }
+
+//---------------------------------------------------------
+//   colorBox
+//---------------------------------------------------------
+
+std::vector<QString> Timeline::getLabels()
+      {
+      QList<Part*> pl = _score->parts();
+      //transfer them into a vector of qstrings and then add the meta row names
+      std::vector<QString> labels;
+      labels.push_back(tr("Tempo"));
+      labels.push_back(tr("Time Signature"));
+      labels.push_back(tr("Key Signature"));
+      labels.push_back(tr("Rehearsal Mark"));
+      labels.push_back(tr("Measures"));
+      for (int r = 0; r < nstaves(); r++) {
+            QTextDocument doc;
+            doc.setHtml(pl.at(correctPart(r))->longName());
```

Comment:
Well user might deliberately use a long name that is completely different from the instrument name.  I would say to use long name if it is non-empty, and then use instrument name as a fall back.